### PR TITLE
NO-ISSUE: Add docs/enhancements directory

### PR DESCRIPTION
**Pull request by djzager**
A slimmed down version of the template one would find in https://github.com/openshift/enhancements is included and referenced.

### CODE REVIEW

File: docs/enhancements/README.md
Comment (mhrivnak) [inline docs/enhancements/README.md:1]: I suggest adding a section addressing when to create an OpenShift enhancement proposal instead. Here are some indicators that could be included in a "rough heuristic":
* requires changes in OpenShift
* requires changes in and/or consensus with other components related to OpenShift
* substantially impacts the requirements for and/or experience of installing and provisioning OpenShift
* would benefit from approval by OpenShift architects

**Review by mhrivnak — CHANGES_REQUESTED** [review]
I like this a lot. There are many design decisions that happen today through a combination of google docs, jira issues, slack, email, etc. that would benefit from a more organized, visible and open process. And as noted by @djzager previously, today it's unclear where someone should start if they want to propose a change. It will also be valuable to have a centralized record of what decisions were made and why.

@avishayt @markmc @romfreiman @ronniel1

**Comment by avishayt** [comment]
This is a great initiative, thanks @djzager !  The size of this project and the number of contributors has grown, and this will certainly help it continue to scale.
I think we also want a section for UI impact.

File: docs/enhancements/template.md
Comment (empovit) [inline docs/enhancements/template.md:58]: Isn't it a chicken-and-egg situation? That is, an enhancement proposal requires design details, a test plan, etc. which probably don't make sense to work on until/unless the proposal has been approved. A few OpenShift enhancement documents I saw looked pretty much like a detailed design. Do we want it here? I'm also afraid many people on the team won't be happy with that.

File: docs/enhancements/template.md
Comment (carbonin) [inline docs/enhancements/template.md:13]: Do these have some sort of "official" definition?
For example, what's the gating criteria for moving from `provisional` (which I'm assuming to be WIP) to `implementable`?

What would we use `informational` for? Wouldn't that just be documentation? Or would we record information about a direction we didn't want to go?

File: docs/enhancements/template.md
Comment (carbonin) [inline docs/enhancements/template.md:19]: Is this supposed to be the inverse of `replaces`?
If that's the case, then shouldn't the doc here be something that happens after this doc? The names are just a bit confusing.

File: docs/enhancements/template.md
Comment (carbonin) [reply docs/enhancements/template.md:58]: If I understand correctly, these are not prerequisites to starting development on this feature, but things that need to be sorted out before release.

That said, this feels like it belongs in JIRA rather than our enhancement doc especially because it includes things that other teams (who may not know or care about the enhancement docs here) will need to complete. For example, test plan (QE team), operational readiness (app-sre?), documentation all feel like we would might need to create separate tickets for those teams

File: docs/enhancements/template.md
Comment (carbonin) [inline docs/enhancements/template.md:136]: This doesn't feel relevant to us.

File: docs/enhancements/template.md
Comment (carbonin) [inline docs/enhancements/template.md:131]: As we're deploying as a SaaS as well as releasing I'm not sure this makes sense.

File: docs/enhancements/template.md
Comment (carbonin) [inline docs/enhancements/template.md:129]: Isn't there a whole separate system QE uses for documenting test plans?
Is this more of a section that would give them hints about specific cases to be sure to test rather than an exhaustive plan?

File: docs/enhancements/template.md
Comment (ronniel1) [reply docs/enhancements/template.md:58]: I think that we said that we'll start by handling just our designs in git for now.
Maybe we should start with a lighter template that is more geared towards design documents?

**Comment by djzager** [comment]
/hold

It's clear based on discussion with @avishayt, @ronniel1, @carbonin, and others that there is debate about whether to pursue this in its current form. I'll work to address comments if/once that question is answered.

**Comment by carbonin** [comment]
> Could you write down, in this PR, what your concerns with this process are? 

Mainly I want to ensure that we are only adding a single document to the process. There was mention of an "enhancement proposal" as well as a "design document". I think one doc can and should serve both purposes. The workflow we use can be different depending on where the idea originates from (maybe there's an epic first, maybe the doc comes first).

If we have a separate design doc I suspect that the enhancement proposal process and template will end up going unused as the vast majority of development is still driven by Red Hat product management.

Additionally, my review was mostly focused on keeping the objectives of this new document (design proposal? enhancement doc? ... :man_shrugging: ) distinct from what is accomplished with an issue tracker. IMO we should focus on the problem we're trying to solve, the chosen (or proposed) solution, pros/cons, and corner cases. I think we specifically should not be talking about test plans, documentation, and ops much at all. Internally those are generally the responsibility of other teams or require internal collaboration to deal with, an external contributor wouldn't really know about those processes unless we move all of that upstream.

File: docs/enhancements/template.md
Comment (djzager) [reply docs/enhancements/template.md:19]: removing this to simplify.

File: docs/enhancements/template.md
Comment (djzager) [reply docs/enhancements/template.md:13]: I'm dropping status as I believe either it is accepted as implementable (merged) or it is rejected (closed). If we need more catagories later...we can add them when it is appropriate.

File: docs/enhancements/template.md
Comment (djzager) [reply docs/enhancements/template.md:58]: Dropping this entirely.

File: docs/enhancements/template.md
Comment (djzager) [reply docs/enhancements/template.md:103]: I'm leaving the test plan section as I think this is more specifically about the e2e and integration tests that we would write here, and in assisted-test-infra.

My belief is that the QE folks would focus on the user stories themselves to guide the tests they write (ie. how is it supposed to be used?).

File: docs/enhancements/template.md
Comment (djzager) [reply docs/enhancements/template.md]: dropping the release signoff as I think it makes the process to prescriptive.

**Comment by djzager** [comment]
I spent some time reviewing the "Design Doc Template" that @empovit has been working on trying to find a middle ground where enhancements can cover simple things as well as complex requests that _need_ detailed designs in order to have meaningful discussion.

> Mainly I want to ensure that we are only adding a single document to the process. 

I wholeheartedly agree with this. My feeling is that the template now leaves room for more information to be added (like fully explaining the design of the enhancement) without prescribing the contributor must include one.

Additionally, I removed the status and release signoff sections as I think the introduce a level of complexity that we just aren't ready for yet. I said it in a comment but I will reiterate how I think it would work:
1. contributor submits an enhancement PR
2. we accept it (it gets merged) implies that we think it can and should be implemented
3. we reject it (it gets closed) implies that we will not implement the enhancement

@avishayt had mentioned:
> For design documents I would like to see:
Design (include any new API definitions)
Upgrade considerations
Security considerations
Metrics
Events
Service interactions and failure handling

As I see it, these all could fall under some combination of User Stories, Implementation Details, Risks and Mitigations, Design Details, and maybe even Drawbacks. That is I believe that the nature of the enhancement will dictate whether a discussion of upgrade considerations belongs in design details (maybe it's specifically concerned with upgrade) or risks and mitigations.

My hope is that we can have one landing pad for enhancement requests to the Assisted Installer family of projects that invites others to contribute out in the open and influence the direction of the project as a whole.

Let me know what you think.

File: docs/enhancements/template.md
Comment (carbonin) [reply docs/enhancements/template.md:114]: :+1: sounds good to me

**Comment by carbonin** [comment]
Looks great! Thanks for putting the work into revising this @djzager.

I'll wait for others input before actually giving it the merge labels though.
Also .. do we have a way to avoid running all the test suites for things like docs?

**Comment by djzager** [comment]
> Also .. do we have a way to avoid running all the test suites for things like docs?

probably something in the Jenkinsfile + additional config changes in openshift/release

**Comment by flaper87** [comment]
Thanks for incorporating the work that @empovit did elsewhere. I like this proposal and the template. It's flexible enough for people to keep it short or go in more length if needed.

At this point, I would love to hear @empovit @avishayt and @ronniel1 thoughts on the current version, see if we can make progress on the idea.

File: docs/enhancements/template.md
Comment (empovit) [inline docs/enhancements/template.md:67]: I think this could benefit from a brief example, just like you did in `Open Questions`. Especially that user stories usually have a particular structure.

**Comment by empovit** [comment]
Looks great! Thanks for the hard work!

What I'm missing though is a list of projects that more likely than not will be affected by any enhancement. At least as a checklist. There is a `UI Impact`, but no mentioning of `Test Infra`, for instance. Test infra is particularly easy to overlook. IMO such a list could help members of the general public in the first place, but also the team members, to understand the real impact of an enhancement.

* Assisted Service
* Assisted Installer & Controller
* Assisted Installer Agent
* Test Infra

Other sections that I'd like to see, again similarly to `UI Impact`, are `API Changes` and `Makefile Changes`. They are special because of the impact they have on the developers. 

I believe both points mentioned above would go into the design section. The projects part can be probably a slightly changed version of https://github.com/openshift/assisted-service/pull/1227/files#diff-8becbf56d4b1acae6a3273a67ef06ab0e7f33967956e4cf8c0ec6486940c51fbR77, so that the total length of the template doesn't change. You've succeeded in keeping it concise :)

**Comment by flaper87** [comment]
> IMO such a list could help members of the general public in the first place, but also the team members, to understand the real impact of an enhancement.

I like this suggestion and the `test-infra` example is a good one. Generally, every project in the assisted-installer ecosystem should leverage this enhancement process and having a list would help to provide insights on the existing ones. In the future, we should document this list of projects somewhere else and just reference it here.

**Comment by avishayt** [comment]
Overall looks good to me.  We can start with this and improve as necessary.
@romfreiman and @ronniel1 should provide input.

**Comment by avishayt** [comment]
Maybe my only concern is things that overlap with the JIRA epic, like user stories

**Comment by flaper87** [comment]
@romfreiman @ronniel1 @celebdor  would be great to have your :+1: / :-1:  before moving forward with this! :pray:

**Comment by romfreiman** [comment]
I like the transparency, I do have few concerns:
1. Will it make us slower?
2. Jira vs the enhancement?
3. Who's in charge that people will write enhancements? We barely can make you update jiras status.

In general, Im not against, just want to be sure that we keep on working in a good pace and we will NEVER EVER use a non-merged enhancement as an excuse for a delay. If I get ack on this, lets proceed.

**Comment by carbonin** [comment]
> we will NEVER EVER use a non-merged enhancement as an excuse for a delay

I'll ask again ...
Can we somehow disable ci on the enhancements and docs directory? Or maybe have a policy that we can merge these manually (via GitHub merge button) rather than waiting for the automation to do it?

**Comment by romfreiman** [comment]
> > we will NEVER EVER use a non-merged enhancement as an excuse for a delay
> 
> I'll ask again ...
> Can we somehow disable ci on the enhancements and docs directory? Or maybe have a policy that we can merge these manually (via GitHub merge button) rather than waiting for the automation to do it?

We can merge manually, yes. I'm more concerned about the state of mind rather than CI failures though :)

**Comment by lranjbar** [comment]
> > we will NEVER EVER use a non-merged enhancement as an excuse for a delay
> 
> I'll ask again ...
> Can we somehow disable ci on the enhancements and docs directory? Or maybe have a policy that we can merge these manually (via GitHub merge button) rather than waiting for the automation to do it?

We can have the docs be ignored. I looked into it, there are some workarounds available.

**Comment by celebdor** [comment]
> I like the transparency, I do have few concerns:
> 
>     1. Will it make us slower?

I don't think so. The time spent doing this will be recovered by all the people being able to look at it and benefit from referencing it while working on the implementation, reducing syncing effort with documentation, people getting up to speed with the project, etc.

> 
>     2. Jira vs the enhancement?

The enhancement fleshes out bits of functionality and design in the open source repository. The content In Jira is related in that  one usually finds there the breakdown of tasks, which contain bits and pieces of the design, but they are harder to consume and piece together.

> 
>     3. Who's in charge that people will write enhancements? We barely can make you update jiras status.

The usual process is that it falls on the reviewers to suggest that something needs an enhancement due to novelty or lack of coverage of current significant work by the current enhancements.

> 
> 
> In general, Im not against, just want to be sure that we keep on working in a good pace and we will NEVER EVER use a non-merged enhancement as an excuse for a delay. If I get ack on this, lets proceed.

This should not make people feel blocked. We often implement and present OpenShift enhancement proposals in parallel. What can happen sometimes is that if the enhancement lags behind the implementation, we may end up having to go back and apply some of the things learned/improved in the enhancement discussion to the implementation.

I understand that when there's an enhancement process sometimes one can feel the will to fully converge before getting their hands dirty with implementation. Maybe an improvement should be that the template asks if you already PoCed or have a working design as one of the check-list items, to encourage that behavior.

**Comment by flaper87** [comment]
I think @celebdor put it very well in his last comment so I won't try to put more words to what he said but I do want to explicitly say that I agree 100% with the comment.

> We can have the docs be ignored. I looked into it, there are some workarounds available.

@lisa is this something you reckon we could do in parallel to this PR? 

From what I can tell, we can move forward with this PR... unless someone has other concerns that should be brought up.

**Comment by lisa** [comment]
> I think @celebdor put it very well in his last comment so I won't try to put more words to what he said but I do want to explicitly say that I agree 100% with the comment.
> 
> > We can have the docs be ignored. I looked into it, there are some workarounds available.
> 
> @lisa is this something you reckon we could do in parallel to this PR?
> 
> From what I can tell, we can move forward with this PR... unless someone has other concerns that should be brought up.

Do you mean a different Lisa?

**Comment by lranjbar** [comment]
> > I think @celebdor put it very well in his last comment so I won't try to put more words to what he said but I do want to explicitly say that I agree 100% with the comment.
> > > We can have the docs be ignored. I looked into it, there are some workarounds available.
> > 
> > 
> > @lisa is this something you reckon we could do in parallel to this PR?
> > From what I can tell, we can move forward with this PR... unless someone has other concerns that should be brought up.
> 
> Do you mean a different Lisa?

@lisa Hi Lisa. I think he probably meant me. Awesome tag and Happy Friday! 🥳 

@flaper87 We've limited the long running jobs with openshift/release#17359 The only downside is that for jobs that don't have `always_run: true` they lose the required tag in Github. This would be a reason to get a docs repo, but otherwise I think we should merge this. :)

**Comment by flaper87** [comment]
I think https://github.com/openshift/assisted-service/pull/1508 could have used an enhancement proposal where the API is defined and discussed. It's unclear, from the PR and the associated issue, where the discussion actually happened, how the API was designed, what guarantees it should provide, and what the actual requirements are.

Can we please move forward with this process? It's not meant to be used for everything but, I do hope that we can encourage people to write more detailed proposals for enhancements like the one mentioned above. 

@ronniel1 @romfreiman @celebdor  I think we are all waiting for you to give a final :+1: / :-1:

**Comment by flaper87** [comment]
I'm approving this based on @romfreiman and @celebdor :+1:s on my previous comment. 

DeepinScreenshot_select-area_20210421113134

/lgtm

**Review by mhrivnak — APPROVED** [review]

**Review by flaper87 — APPROVED** [review]

**Comment by osherdp** [comment]
/hold
@djzager you might need to rebase in order to make build pass, you're on top of a very old commit :)

**Review by avishayt — APPROVED** [review]
Let's get this moving and adjust as necessary

**Pull request merged**
